{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2755844",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# Import GenoTools library for Wrayner tool\n",
    "from genotools.utils import shell_do, merge_genos\n",
    "from genotools.imputation import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9859f4ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Release of interest\n",
    "rel = 10\n",
    "\n",
    "# Create folders to mount\n",
    "! mkdir releases release ref_files working\n",
    "\n",
    "# Read Access only\n",
    "! gcsfuse --dir-mode 555 --file-mode 444 --implicit-dirs gp2tier2_vwb releases\n",
    "! gcsfuse --dir-mode 555 --file-mode 444 --implicit-dirs gp2_release{rel} release\n",
    "! gcsfuse --dir-mode 555 --file-mode 444 --implicit-dirs gtserver-eu-west4-gp2-release-terra ref_files\n",
    "\n",
    "# Read/Write Access\n",
    "! gcsfuse --dir-mode 777 --file-mode 777 --implicit-dirs gp2_working_eu working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c7ed383",
   "metadata": {},
   "outputs": [],
   "source": [
    "wd = '/YOUR/WORKING/DIR'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc963bee",
   "metadata": {},
   "source": [
    "### Create Files with Sex Info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aabd7fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep subset of IDs based on sex\n",
    "label = 'AJ'\n",
    "user = 'nicole'\n",
    "sex = 'male'\n",
    "\n",
    "! mkdir {wd}/working/{user}/x_chrom_new/imputation/{label}\n",
    "\n",
    "master = pd.read_csv(f'{wd}/releases/release{rel}/clinical_data/master_key_release{rel}_final_vwb.csv')\n",
    "ancestry = master[(master.nba_label == label) & (master.biological_sex_for_qc == sex.title())]\n",
    "ancestry['FID'] = 0\n",
    "ancestry[['FID', 'GP2ID']].to_csv(f'{wd}/working/{user}/x_chrom_new/imputation/{label}/{label}_release{rel}_{sex}.samples', sep = '\\t', index = False, header = None)\n",
    "\n",
    "# Check if any samples were pruned\n",
    "display(ancestry.nba_prune_reason.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18d08660",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For EUR: Split key into 3 parts due to sample limits on impuation\n",
    "label = 'EUR'\n",
    "sex = 'male'\n",
    "\n",
    "# Shuffle the DataFrame within each phenotype group\n",
    "ancestry = master[(master.nba_label == label) & (master.biological_sex_for_qc == sex.title())]\n",
    "df = ancestry.groupby('GP2_phenotype_for_qc', group_keys=False).apply(lambda x: x.sample(frac=1, random_state=42)).reset_index(drop=True)\n",
    "\n",
    "# Split each phenotype group into three equal parts\n",
    "grouped = df.groupby('GP2_phenotype_for_qc')\n",
    "df1 = pd.concat([grp.iloc[:len(grp) // 3] for _, grp in grouped])\n",
    "df2 = pd.concat([grp.iloc[len(grp) // 3:2 * len(grp) // 3] for _, grp in grouped])\n",
    "df3 = pd.concat([grp.iloc[2 * len(grp) // 3:] for _, grp in grouped])\n",
    "\n",
    "df1['FID'] = 0\n",
    "df2['FID'] = 0\n",
    "df3['FID'] = 0\n",
    "\n",
    "# Check phenotype counts in each part\n",
    "print(\"Part 1 counts:\\n\", len(df1), df1['GP2_phenotype_for_qc'].value_counts(), '\\n')\n",
    "print(\"Part 2 counts:\\n\", len(df2), df2['GP2_phenotype_for_qc'].value_counts(), '\\n')\n",
    "print(\"Part 3 counts:\\n\", len(df3), df3['GP2_phenotype_for_qc'].value_counts())\n",
    "\n",
    "df1[['FID', 'GP2ID']].to_csv(f'{wd}/working/{user}/x_chrom_new/imputation/EUR_part1/EUR_part1_release{rel}_{sex}.samples', sep = '\\t', index = False, header = None)\n",
    "df2[['FID', 'GP2ID']].to_csv(f'{wd}/working/{user}/x_chrom_new/imputation/EUR_part2/EUR_part2_release{rel}_{sex}.samples', sep = '\\t', index = False, header = None)\n",
    "df3[['FID', 'GP2ID']].to_csv(f'{wd}/working/{user}/x_chrom_new/imputation/EUR_part3/EUR_part3_release{rel}_{sex}.samples', sep = '\\t', index = False, header = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "986c8676",
   "metadata": {},
   "source": [
    "### Explore Release-Wide GenoTools Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6ddf773",
   "metadata": {},
   "outputs": [],
   "source": [
    "# EUR run can begin here because ID subsets were pre-made -- ADD CODE FOR SPLITTING EUR INTO 3 CHUNKS\n",
    "user = 'nicole'\n",
    "sex = 'female'\n",
    "group = f'r{rel}_final'\n",
    "label = 'EUR_part1' # Can also choose: EUR_part2, EUR_part3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64be5ad8",
   "metadata": {},
   "outputs": [],
   "source": [
    "keep_file = f'{wd}/working/{user}/x_chrom_new/imputation/{label}/{label}_release{rel}_{sex}.samples'\n",
    "\n",
    "# Check that XY and X Regions Exist\n",
    "test = pd.read_csv(f'{wd}/release/{group}/GP2_{group}_post_genotools_{label}.pvar', sep = '\\s+', low_memory = False)\n",
    "df2 = pd.DataFrame(test['#CHROM'].value_counts()).reset_index()\n",
    "chrom_list = ['X', 'Y', 'XY', 'MT']\n",
    "df2 = df2[df2['#CHROM'].isin(chrom_list)]\n",
    "display(df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7876d58f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split into sex of interest\n",
    "input_file = f'{wd}/release/{group}/GP2_{group}_post_genotools_{label}'\n",
    "output_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}'\n",
    "\n",
    "# Change directory to /dev/shm for more stoage + easy cleanup for tmp files\n",
    "os.chdir('/dev/shm')\n",
    "\n",
    "! mkdir imputation imputation/{group}\n",
    "! plink2 --pfile {input_file} --keep {keep_file} --make-pgen --out {output_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8699c1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Double check that output is one sex\n",
    "check_bim = pd.read_csv(f'{output_file}.psam', sep = '\\t')\n",
    "display(check_bim.SEX.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ebcdf14",
   "metadata": {},
   "source": [
    "### Region-Specific QC for PAR and Non-PAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68e84d92",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = 'PAR'\n",
    "\n",
    "input_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}'\n",
    "output_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051323db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset by region\n",
    "if region == 'non_PAR':\n",
    "    ! plink2 --pfile {input_file} --chr X --set-invalid-haploid-missing --make-pgen --out {output_file}\n",
    "elif region == 'PAR':\n",
    "    ! plink2 --pfile {input_file} --chr XY --make-pgen --out {output_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4587b303",
   "metadata": {},
   "outputs": [],
   "source": [
    "if region == 'PAR':\n",
    "    # Load the .bim file\n",
    "    pvar_file = f'{output_file}.pvar'\n",
    "    pvar = pd.read_csv(pvar_file, sep=\"\\t\", low_memory = False)\n",
    "\n",
    "    # Replace chromosome names\n",
    "    pvar['#CHROM'] = pvar['#CHROM'].replace({'XY': 'X'})\n",
    "    display(pvar['#CHROM'].value_counts())\n",
    "\n",
    "    # Save the updated .bim file\n",
    "    pvar.to_csv(f'{output_file}.pvar', sep=\"\\t\", index=False)\n",
    "\n",
    "    # Split-PAR and remove any remaining X\n",
    "    input_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}'\n",
    "    output_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}_only'\n",
    "\n",
    "    ! plink2 --pfile {input_file} --split-par b38 --make-pgen --out {output_file}\n",
    "\n",
    "    # Only keep PAR1 and PAR2\n",
    "    input_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}_only'\n",
    "    output_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}_only_fixed'\n",
    "\n",
    "    ! plink2 --pfile {input_file} --chr PAR1,PAR2 --make-pgen --out {output_file}\n",
    "\n",
    "    # Load the .bim file again\n",
    "    pvar_file = f'{output_file}.pvar'\n",
    "    pvar = pd.read_csv(pvar_file, sep=\"\\t\", low_memory = False)\n",
    "\n",
    "    # Replace chromosome names to avoid removal during imputation\n",
    "    display(pvar['#CHROM'].value_counts())\n",
    "    pvar['#CHROM'] = pvar['#CHROM'].replace({'PAR1': 'X', 'PAR2': 'X'})\n",
    "    display(pvar['#CHROM'].value_counts())\n",
    "\n",
    "    # Save the updated .bim file\n",
    "    pvar.to_csv(f'{output_file}.pvar', sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "727d6d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finalize PLINK files per region\n",
    "if region == 'PAR':\n",
    "    input_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}_only_fixed'\n",
    "elif region == 'non_PAR':\n",
    "    input_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}'\n",
    "output_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_{region}_final'\n",
    "\n",
    "! plink2 --pfile {input_file} --output-chr 26 --make-bed --out {output_file}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05ce63",
   "metadata": {},
   "source": [
    "### Save files to permanent storage in Google Buckets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a984bff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "if sex == 'female':\n",
    "    par_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_PAR_final'\n",
    "    non_par_file = f'/dev/shm/imputation/{group}/GP2_{group}_post_genotools_{label}_{sex}_non_PAR_final'\n",
    "\n",
    "    # Check if they exist\n",
    "    if os.path.exists(f'{par_file}.bed') and os.path.exists(f'{non_par_file}.bed'):\n",
    "        print(\"Both PAR and non-PAR files exist for females! Ready to merge.\\n\")\n",
    "    elif os.path.exists(f'{par_file}.bed'):\n",
    "        print(\"Only PAR file exists for females. Please prepare female non-PAR file before moving forward.\\n\")\n",
    "    elif os.path.exists(f'{non_par_file}.bed'):\n",
    "        print(\"Only Non-PAR file exists for females. Please prepare female PAR file before moving forward.\\n\")\n",
    "    else:\n",
    "        print(\"Neither file exists.\")\n",
    "\n",
    "    # Merge PAR and non-PAR regions for females\n",
    "    output_file = f'GP2_{group}_post_genotools_{label}_{sex}_final'\n",
    "    ! plink --bfile {non_par_file} --bmerge {par_file} --make-bed --out {output_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98a3e054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy over final output to Bucket storage\n",
    "! mkdir {wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}\n",
    "\n",
    "if sex == 'male':\n",
    "    ! mkdir {wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}/{region}\n",
    "    ! scp {output_file}.* {wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}/{region}\n",
    "elif sex == 'female':\n",
    "    if not os.path.exists(f'{output_file}.bed'):\n",
    "        print(\"Please make sure that both PAR and Non-PAR files are created and merged for females.\")\n",
    "    else:\n",
    "        ! scp {output_file}.* {wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9bf972b",
   "metadata": {},
   "source": [
    "### Submit to Wrayner Tool for VCF Creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887ca6d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ancestry_list = [label]\n",
    "# ancestry_list = ['EUR_part1', 'EUR_part2', 'EUR_part3'] # if the above sections were completed for all 3 parts of EUR\n",
    "\n",
    "impute_genos_list = []\n",
    "impute_labels_list = []\n",
    "for label in ancestry_list:\n",
    "    if sex == 'male':\n",
    "        path_to_file = f'{wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}/{region}/GP2_{group}_post_genotools_{label}_{sex}_{region}_final'\n",
    "    elif sex == 'female':\n",
    "        path_to_file = f'{wd}/working/{user}/x_chrom_new/imputation/{label}/{sex}/GP2_{group}_post_genotools_{label}_{sex}_final'\n",
    "    impute_genos_list.append(path_to_file)\n",
    "    impute_labels_list.append(label)\n",
    "print(impute_genos_list)\n",
    "print(impute_labels_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "506f9b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TOPMed Preparation\n",
    "ref_dir = f'{wd}/ref_files/clinical'\n",
    "\n",
    "topmed_ref_panel= f'{ref_dir}/PASS.Variants.TOPMed_freeze5_hg38_dbSNP.tab.gz'\n",
    "check_bim_pl = f'{ref_dir}/HRC-1000G-check-bim.pl'\n",
    "\n",
    "for geno, label in zip(impute_genos_list, impute_labels_list):   \n",
    "    if sex == 'male':\n",
    "        imputed_out_dir = f'{wd}/working/nicole/x_chrom_new/imputation/imputation_out_dir/{sex}/{region}/'\n",
    "        imputation_prep = f'{wd}/working/nicole/x_chrom_new/imputation/imputation_prep/{sex}/{region}/'\n",
    "    elif sex == 'female':\n",
    "        imputed_out_dir = f'{wd}/working/nicole/x_chrom_new/imputation/imputation_out_dir/{sex}/'\n",
    "        imputation_prep = f'{wd}/working/nicole/x_chrom_new/imputation/imputation_prep/{sex}/'\n",
    "\n",
    "    label_outdir = f'{imputed_out_dir}/{label}'\n",
    "    \n",
    "    impute_prep_outdir = f'{imputation_prep}/{label}'\n",
    "    impute_prep_geno = f'{impute_prep_outdir}/{label}'\n",
    "\n",
    "    os.makedirs(f'{label_outdir}', exist_ok=True)\n",
    "    os.makedirs(f'{impute_prep_outdir}', exist_ok=True)\n",
    "    \n",
    "    impute_data = impute_data_prep(geno, impute_prep_geno, topmed_ref_panel, check_bim_pl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01151fd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that files were made\n",
    "for label in ancestry_list:\n",
    "    ! ls -lh {imputation_prep}/{label}/{label}_pre_impute_chr{23}.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98faece1",
   "metadata": {},
   "source": [
    "### VCF Prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2427f38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write a chr map file to rename Chr23 to X\n",
    "with open(f\"{wd}/chr_map.txt\", \"w\") as f:\n",
    "    f.write(\"chr23 chrX\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb69ed76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adjust Chromosome 23\n",
    "for label in ancestry_list:\n",
    "      out_folder = f'{imputation_prep}/{label}/'\n",
    "      ! tabix -p vcf {out_folder}/{label}_pre_impute_chr23.vcf.gz -f\n",
    "\n",
    "      # Rename\n",
    "      ! bcftools annotate --rename-chrs {wd}/chr_map.txt -Oz -o {out_folder}/{label}_23_output.vcf.gz {out_folder}/{label}_pre_impute_chr23.vcf.gz\n",
    "      \n",
    "      # Normalize\n",
    "      ! bcftools norm --check-ref s --fasta-ref {wd}/ref_files/clinical/hg38.fa {out_folder}/{label}_23_output.vcf.gz -Oz -o {out_folder}/{label}_pre_impute_chr23_fixed.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e11d0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Double check that VCF files exist to submit to imputation\n",
    "for label in ancestry_list:\n",
    "    ! ls -lh {imputation_prep}/{label}/{label}_pre_impute_chr23_fixed.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03c49e21",
   "metadata": {},
   "source": [
    "### Submit VCF to Imputation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5729b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Double check commands to send to imputation\n",
    "\n",
    "for label in ancestry_list:\n",
    "    if sex == 'female':\n",
    "        region = 'full'\n",
    "    impute_cmd = f'imputationbot impute --files {imputation_prep}/{label}/{label}_pre_impute_chr23_fixed.vcf.gz --refpanel topmed-r3 --build hg38 --r2Filter 0.3 --population all --name x_chrom_{label}_{sex}_{region}'\n",
    "    print(impute_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d08ffe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Submit job\n",
    "\n",
    "for label in ancestry_list:\n",
    "    ! {impute_cmd}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python_3_11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "undefined.undefined.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
