{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c66874e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "286f323e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Release of interest\n",
    "rel = 10\n",
    "\n",
    "# Create folders to mount\n",
    "! mkdir releases release working\n",
    "\n",
    "# Read Access only\n",
    "! gcsfuse --dir-mode 555 --file-mode 444 --implicit-dirs gp2tier2_vwb releases\n",
    "! gcsfuse --dir-mode 555 --file-mode 444 --implicit-dirs gp2_release{rel} release\n",
    "\n",
    "# Read/Write Access\n",
    "! gcsfuse --dir-mode 777 --file-mode 777 --implicit-dirs gp2_working_eu working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dc59d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "wd = '/YOUR/WORKING/DIR'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b408f0d9",
   "metadata": {},
   "source": [
    "### PLINK File Prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "162ef4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize variant names\n",
    "\n",
    "# ! wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz\n",
    "user = 'nicole'\n",
    "label = 'AJ'\n",
    "rel = 10\n",
    "region = 'non_PAR'\n",
    "\n",
    "if region == 'non_PAR': \n",
    "    chrom = 'X'\n",
    "else:\n",
    "    chrom = 'PAR1, PAR2'\n",
    "\n",
    "# Split females into regions for separate PAR and non-PAR analyses\n",
    "female = f'{wd}/working/{user}/x_chrom_new/imputation/imputation_out_dir/{label}/female/chrX_{label}_rel{rel}_female'\n",
    "female_out_path = f'{wd}/working/{user}/x_chrom_new/imputation/imputation_out_dir/{label}/female/{region}/chrX_{label}_rel{rel}_female_{region}'\n",
    "\n",
    "os.chdir('/dev/shm')\n",
    "! plink2 --pfile {female} --chr {chrom} --make-pgen --out {female_out_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2126c694",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run normalization\n",
    "sex = 'female'\n",
    "input = f'{wd}/working/{user}/x_chrom_new/imputation/imputation_out_dir/{label}/{sex}/{region}/chrX_{label}_rel{rel}_{sex}'\n",
    "\n",
    "os.chdir('/dev/shm')\n",
    "! plink2 --pfile {female_out_path} --normalize --fa {wd}/working/{user}/XWAS/hg38.fa.gz \\\n",
    "    --set-all-var-ids \"chr@:#:\\$r:\\$a\" \\\n",
    "    --new-id-max-allele-len 999 --sort-vars \\\n",
    "    --make-pgen --out {wd}/working/{user}/XWAS/{label}/{label}_{region}_{sex}_normalized"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6c44997",
   "metadata": {},
   "source": [
    "### PCs Prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "823eb4a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "rel = 10\n",
    "label = 'AJ'\n",
    "\n",
    "# Using saved release-wide autosomal PCA results -- ADD PCA CALCULATION\n",
    "eigenval_file = f\"{wd}/releases/release{rel}/raw_genotypes/{label}/{label}_release{rel}_vwb.eigenval\" # Male + Female results\n",
    "eigenvals = pd.read_csv(eigenval_file, header=None, names=[\"eigenvalue\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da44d3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate % variance explained and cumulative variance\n",
    "eigenvals[\"variance_explained\"] = eigenvals[\"eigenvalue\"] / eigenvals[\"eigenvalue\"].sum() * 100\n",
    "eigenvals[\"cumulative_variance\"] = eigenvals[\"variance_explained\"].cumsum()\n",
    "\n",
    "# Find the first PC where cumulative variance >= 90\n",
    "pc90 = (eigenvals[\"cumulative_variance\"] >= 90).idxmax() + 1\n",
    "var90 = eigenvals.loc[pc90 - 1, \"cumulative_variance\"]\n",
    "\n",
    "# Create side-by-side plots\n",
    "fig, axes = plt.subplots(1, 2, figsize=(12, 5), sharex=True)\n",
    "\n",
    "# Left: Variance explained\n",
    "axes[0].plot(range(1, len(eigenvals) + 1), eigenvals[\"variance_explained\"], marker=\"o\")\n",
    "axes[0].set_xlabel(\"Principal Component\")\n",
    "axes[0].set_ylabel(\"Variance Explained (%)\")\n",
    "axes[0].set_title(\"Variance Explained per PC\")\n",
    "axes[0].grid(True, linestyle=\"--\", alpha=0.6)\n",
    "\n",
    "# Right: Cumulative variance\n",
    "axes[1].plot(range(1, len(eigenvals) + 1), eigenvals[\"cumulative_variance\"], marker=\"s\", linestyle=\"--\")\n",
    "axes[1].axhline(y=90, color=\"red\", linestyle=\":\", label=\"90% Threshold\")\n",
    "axes[1].scatter(pc90, var90, color=\"red\", zorder=5)\n",
    "axes[1].text(pc90 + 0.5, var90 - 5, f\"PC{pc90} ~ {var90:.1f}%\", color=\"red\")\n",
    "axes[1].set_xlabel(\"Principal Component\")\n",
    "axes[1].set_ylabel(\"Cumulative Variance (%)\")\n",
    "axes[1].set_title(\"Cumulative Variance Explained\")\n",
    "axes[1].legend()\n",
    "axes[1].grid(True, linestyle=\"--\", alpha=0.6)\n",
    "\n",
    "fig.suptitle(f\"Scree Plot: {label} R{rel}\", fontsize=14, y=1.05)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9678a92",
   "metadata": {},
   "source": [
    "### Covar Prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33f3f254",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import release files\n",
    "total_pheno = pd.read_csv(f'{wd}/releases/release{rel}/imputed_genotypes/{label}/chr1_{label}_release{rel}_vwb.psam', sep='\\t')\n",
    "total_pheno.rename(columns = {'#IID': 'IID'}, inplace = True)\n",
    "display(total_pheno.head())\n",
    "\n",
    "pcs = pd.read_csv(f\"{wd}/releases/release{rel}/raw_genotypes/{label}/{label}_release{rel}_vwb.eigenvec\", sep='\\t')\n",
    "display(pcs.head())\n",
    "\n",
    "change = pd.read_csv(f'{wd}/releases/release{rel}/clinical_data/r{rel}_extended_clinical_data_vwb.csv')\n",
    "change = change[['GP2ID','visit_month', 'age_at_baseline', 'primary_diagnosis', 'last_diagnosis']]\n",
    "display(change.head())\n",
    "\n",
    "related = pd.read_csv(f'{wd}/releases/release{rel}/meta_data/related_samples/{label}_release{rel}_vwb.related')\n",
    "display(related.head())\n",
    "\n",
    "master = pd.read_csv(f'{wd}/releases/release{rel}/clinical_data/master_key_release{rel}_final_vwb.csv')\n",
    "display(master.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d318800",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import XWAS prepped psam files\n",
    "region = 'non_PAR'\n",
    "sex = 'male'\n",
    "\n",
    "input = f'{wd}/working/{user}/XWAS/{label}/{label}_{region}_{sex}_normalized'\n",
    "pheno = pd.read_csv(f'{input}.psam', sep='\\t') \n",
    "\n",
    "pheno['IID'] = pheno[\"#IID\"].str.replace(\"^0_\", \"\", regex=True)\n",
    "pheno = pheno.merge(total_pheno[['IID', 'PHENO1']], on = 'IID', how = 'inner')\n",
    "pheno = pheno[['IID', '#IID', 'SEX', 'PHENO1']]\n",
    "display(pheno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4935ff6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge1 = pd.merge(pheno, pcs, on='IID')\n",
    "display(merge1)\n",
    "\n",
    "no_dup = change.dropna(subset=['last_diagnosis'])\n",
    "no_dup['diagnosis_change'] = no_dup.apply(lambda row: 'No' if row['primary_diagnosis'] == row['last_diagnosis'] else 'Yes', axis=1)\n",
    "d_change = no_dup[no_dup['diagnosis_change']=='Yes']\n",
    "d_change.info()\n",
    "\n",
    "anc_d = pd.merge(d_change, merge1, left_on='GP2ID', right_on='#IID')\n",
    "anc_d.info()\n",
    "\n",
    "diag_remove = anc_d['GP2ID']\n",
    "anc_updated = merge1[~merge1['IID'].isin(diag_remove)]\n",
    "anc_updated.info()\n",
    "\n",
    "remove = related['IID1']\n",
    "merge_norel = anc_updated[~anc_updated['IID'].isin(remove)]\n",
    "merge_norel.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49e8dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_control = master[master['baseline_GP2_phenotype']=='Population Control']\n",
    "pop_control.head()\n",
    "\n",
    "pop_change = pop_control['GP2ID']\n",
    "merge_norel['PHENO1'][merge_norel['IID'].isin(pop_change)] = 1\n",
    "merge_norel.info()\n",
    "\n",
    "merge_pheno = merge_norel[~merge_norel['PHENO1'].isnull()]\n",
    "merge_pheno.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b7768c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.merge(master, merge_pheno, left_on='GP2ID', right_on='IID')\n",
    "test.info()\n",
    "\n",
    "bad = test[(test['study']=='BCM') & (test['PHENO1']==1)]\n",
    "bad.head()\n",
    "\n",
    "test = test[~test['GP2ID'].isin(bad['GP2ID'])]\n",
    "test.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bd7ae33",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge1_final = test[['IID', '#IID', 'PHENO1', 'SEX', 'age_of_onset', 'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10']]\n",
    "merge1_final.columns = ['IID', '#IID', 'PHENO1', 'SEX', 'AGE', 'PC1', 'PC2', 'PC3', 'PC4', 'PC5', 'PC6', 'PC7', 'PC8', 'PC9', 'PC10']\n",
    "merge1_final.head()\n",
    "\n",
    "merge1_final.fillna('-9', inplace=True)\n",
    "\n",
    "pheno_final = merge1_final[['#IID', 'PHENO1']]\n",
    "pheno_final.rename(columns={'PHENO1': 'PHENO'}, inplace = True)\n",
    "\n",
    "merge1_final.drop(columns = ['IID', 'SEX', 'AGE', 'PHENO1'], inplace = True)\n",
    "\n",
    "display(pheno_final)\n",
    "display(merge1_final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3209d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "pheno_final.to_csv(f'{wd}/prep_files/{sex}_{label}_pheno.txt', sep='\\t', index=False)\n",
    "merge1_final.to_csv(f'{wd}/prep_files/{sex}_{label}_covar.txt', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
