{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e3306b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from qmplot import qqplot\n",
    "from src.plotting import load_plink_glm, miami_plot_with_PAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d66aa3a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create folders to mount\n",
    "! mkdir working\n",
    "\n",
    "# Read/Write Access\n",
    "! gcsfuse --dir-mode 777 --file-mode 777 --implicit-dirs gp2_working_eu working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5012aa4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wd = '/YOUR/WORKING/DIR'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6145127",
   "metadata": {},
   "source": [
    "### Miami Plot of GWAS Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c52d4dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot results from both sexes and regions\n",
    "label = 'AJ'\n",
    "user = 'nicole'\n",
    "\n",
    "region = 'non_PAR'\n",
    "fem_nonpar_path = f'{wd}/working/{user}/XWAS/{label}/XWAS_{label}_{region}_females.PHENO.glm.logistic.hybrid'\n",
    "male_nonpar_path= f'{wd}/working/{user}/XWAS/{label}/XWAS_{label}_{region}_males.PHENO.glm.logistic.hybrid'\n",
    "\n",
    "region = 'PAR'\n",
    "fem_par_path = f'{wd}/working/{user}/XWAS/{label}/XWAS_{label}_{region}_females.PHENO.glm.logistic.hybrid'\n",
    "male_par_path= f'{wd}/working/{user}/XWAS/{label}/XWAS_{label}_{region}_males.PHENO.glm.logistic.hybrid'\n",
    "\n",
    "df_f_nonpar = load_plink_glm(fem_nonpar_path)\n",
    "df_f_par = load_plink_glm(fem_par_path)\n",
    "df_m_nonpar = load_plink_glm(male_nonpar_path)\n",
    "df_m_par = load_plink_glm(male_par_path)\n",
    "\n",
    "fig = miami_plot_with_PAR(\n",
    "    male_nonpar=df_m_nonpar,\n",
    "    male_par=df_m_par,\n",
    "    fem_nonpar=df_f_nonpar,\n",
    "    fem_par=df_f_par,\n",
    "    title=f\"{label} X Chromosome Results\",\n",
    "    genomewide=5e-8,\n",
    "    suggestive=1e-6,\n",
    "    nonpar_color=\"#76b7e5\",\n",
    "    par_color=\"#4f77be\",\n",
    "    point_size=5,\n",
    "    annotate_top_n=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d13966",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.savefig(f\"{wd}/{label}_miami_chrX.png\", dpi=600, bbox_inches=\"tight\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e16fb7c1",
   "metadata": {},
   "source": [
    "### Plot QQ Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ccac0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# QQ Plots with all regions and sexes in one figure\n",
    "plot_info = [\n",
    "    (df_f_nonpar['P'], \"Female Non-PAR\"),\n",
    "    (df_f_par['P'],    \"Female PAR\"),\n",
    "    (df_m_nonpar['P'], \"Male Non-PAR\"),\n",
    "    (df_m_par['P'],    \"Male PAR\")\n",
    "]\n",
    "\n",
    "fig, axes = plt.subplots(2, 2, figsize=(18, 12), facecolor=\"w\")\n",
    "\n",
    "for ax, (pdata, title) in zip(axes.ravel(), plot_info):\n",
    "    qqplot(\n",
    "        data=pdata,\n",
    "        marker=\"o\",\n",
    "        title=title,\n",
    "        xlabel=r\"Expected -log(P)\",\n",
    "        ylabel=r\"Observed -log(P)\",\n",
    "        ax=ax\n",
    "    )\n",
    "\n",
    "plt.tight_layout()\n",
    "fig.savefig(f\"{label}_qq_plots.png\", dpi=600, bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
